### DISCUSSION TITLE
The default Value 30m of CACHE_TTL is not Working when shift regions

### DISCUSSION THREAD

**Discussion by SerenaIce** [directus/directus#10241] [Archive]
I have two directus applications deploying in both us-east-1 and us-east-2 of AWS. And I also deployed two elasticache for redis in both regions. Which are connecting to the same database. If I created a new item in directus in us-east-1, then I shifted the route to us-east-2, the new created item was not showing up because the cache in us-east-2 had not been synchronized yet. 

I didn't set the CACHE_AUTO_PURGE for now which might solve this problem. But the documentation says that the default value for CACHE_TTL is 30m. I waited for a long more than 30 minutes, the new created item was not showing up in directus in us-east-2 until I manually did some other items changes, the cache synchronization seemed being triggered.

My questions:
Why the CACHE_TTL doesn't work for this kind of situation?
In this kind of case, if I didnt do any data change, the data would not be synchronized in us-east-2, isn't it?

**Comment by rijkvanzanten** [comment]
> Why the CACHE_TTL doesn't work for this kind of situation?
> In this kind of case, if I didnt do any data change, the data would not be synchronized in us-east-2, isn't it?

I think the answer to both of these questions is that you're effectively dealing with lag in the times based on the distributed nature of the infrastructure. If instance A set the cache in region A, instance B won't clear the cache in region A and vice versa. The TTL is checked by the instance that reads from the cache, so if you were reading from B and looking at cache A, you won't get the results you expect.

Seeing that you're running Directus itself in multiple regions with multiple caches, I would recommend setting the CACHE TTL to something very low, or disabling caches altogether, seeing that you don't have any replication on the ElastiCache layer between the different regions. By having the redis instances out of sync, you might run into various issues in the caches and rate limiting

**Reply by SerenaIce** [reply]
So what does the time setting for CACHE_TTL really mean? Although I created the item in region A and it's cached in cache in region A, the item was stored into the database. So from my understanding, after the CACHE_TTL time, the data should be synchronized from   database to the cache B in region B, right?

And another thing needs to confirm is, if I set the CACHE_AUTO_PURGE to true, should I solve the issue I've encountered?

**Reply by rijkvanzanten** [reply]
> So from my understanding, after the CACHE_TTL time, the data should be synchronized from database to the cache B in region B, right?

No. When Directus 1 (D1) gets the request for an item, it checks if the item exists in its cache (C1), and if the record hasn't expired based on the TTL setting. If the cache exists, the requested record is returned straight from C1. Otherwise, Directus will read the record fresh from the database, and store a copy in the C1 for next time.

If you have two copies of Directus connected to two different Redis clusters, they won't be able to modify eachother's cache. D1 can't modify C2 in any way, as they're physically separated.

This can cause the caches to get out of sync. D1 could read an item and save it to C1. If D2 gets a request for the same item, the cache doesn't exist, as D2 isn't aware of C1, it can only read from C2. This can cause D2 to return a different item than D1, as D2 might have something else cached than D1.

> And another thing needs to confirm is, if I set the CACHE_AUTO_PURGE to true, should I solve the issue I've encountered?

CACHE_AUTO_PURGE will purge the cache whenever a write operation is made. However, the same separation applies. If an item is changed in the database by D1, only C1 will be purged, as D1 isn't aware of C2.

---

Generally speaking, you'd want to hook both D1 and D2 up to the same Redis instance, to make sure they share the same cache 👍🏻

**Reply by SerenaIce** [reply]
Yes. But if TTL expired, C2 also should refresh from DB and then D2 should get the data from C2, right? I mean the purge option will not affect D2 and C2, but the TTL should applied to both. So if the DB is changed, then D2 and C2 will eventually updated after the TTL time. Because for both D1 and D2, they have the default value 30m for CACHE_TTL.

**Reply by rijkvanzanten** [reply]
That's correct. Do keep in mind that the actual time that happens might be different for both caches though, as C1 and C2 might've been populated at different times.

**Reply by SerenaIce** [reply]
So the CACHE_TTL should work, but now for this kind of case it doesn't, no matter how long I waited, the data would not be refreshed automatically. So please identify this as a bug and if you have any further solution, please let me know. Thanks.

**Reply by rijkvanzanten** [reply]
@SerenaIce If there's a bug around the TTL, please [create a new issue](https://github.com/directus/directus/issues/new) outlining the exact steps to reproduce 👍🏻
